### PR TITLE
fix: downtype Pydantic types which cannot be instantiated

### LINKED ISSUE #2334: Bug: Pydantic types that cannot be instantiated - deserialization
### Description

When deserializing object, Litestar checks if a given `value` matches a given `target_type`. Additional `type_decoders` can be provided to convert a type given after deserialization to the expected type.

The problem is - when using Pydantic models and PydanticDTO's - Pydantic has some types that cannot be instantiated and are really not supposed to be used elsewhere other than model declaration - for example `EmailStr`.

When deserializing an object like:
```
{
  "email": "user@example.com"
}
```
using a DTO given in the MCVE -> the expected type of `email` during runtime is `EmailStr` (see screenshot). Msgspec after deserialization returns a `str` so a decoder is needed. However, [there is no way to instantiate `EmailStr`](https://github.com/pydantic/pydantic/blob/main/pydantic/networks.py#L171) - it is only meant to be used as a type annotation for a Pydantic model. 
`cast` doesn't change the type at runtime it is only useful for type checkers. So such decoder is impossible.

The solution should be changing the `target_type` to `str` in the `default_deserializer` function - [line 91 of `litestar/serialization/msgspec_hooks.py`](https://github.com/litestar-org/litestar/blob/main/litestar/serialization/msgspec_hooks.py#L91) and consequently in the call to [`msgspec.json.decode` at line 187 of the same file](https://github.com/litestar-org/litestar/blob/main/litestar/serialization/msgspec_hooks.py#L187). Internally by Pydantic, they're treated as a `str` after validation is performed.

Which should be handled by the PydanticDTO. Also there's more types in Pydantic that cannot be instantiated like `ImportStr` and maybe others. The target type for these should always be they're naive counterpart.

### URL to code causing the issue

_No response_

### MCVE

```python
import logging

from litestar import Litestar, post
from litestar.contrib.pydantic import PydanticDTO
from litestar.dto import DTOConfig
from pydantic import BaseModel, EmailStr

from typing import cast


class EmailModel(BaseModel):
    email: EmailStr


class EmailDTO(PydanticDTO[EmailModel]):
    config = DTOConfig()


@post("/email", dto=EmailDTO)
async def accept_email(data: EmailModel) -> None:
    logging.info(data)


def is_emailstr_type(obj_type: type[EmailStr]) -> bool:
    return obj_type is EmailStr


def emailstr_decoder(obj_type: type[EmailStr], value: str) -> EmailStr:
    # How?
    
    return cast(EmailStr, value)


app = Litestar(
    route_handlers=[accept_email], type_decoders=[(is_emailstr_type, emailstr_decoder)]
)
```

### Steps to reproduce

1. Run the Litestar app from the MCVE
2. Send POST request with the body of `{"email": "user@example.com"}` to `/email`
3. `msgspec.ValidationError` occurs
4. Also when removing `type_decoders`, a similar error appears

### Screenshots

Expected type cannot be instantiated

### Logs
<details>
<summary>Full traceback</summary>

```
INFO:     127.0.0.1:58390 - "POST /email HTTP/1.1" 400 Bad Request
ERROR - 2023-09-22 21:00:43,846 - litestar - config - exception raised on http connection to route /email

Traceback (most recent call last):
Traceback (most recent call last):
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/serialization/msgspec_hooks.py", line 187, in decode_json
    return msgspec.json.decode(
           ^^^^^^^^^^^^^^^^^^^^
msgspec.ValidationError: decoding to str: need a bytes-like object, type found - at `$.email`

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 184, in _get_response_data
    kwargs["data"] = await kwargs["data"]
                     ^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/_kwargs/extractors.py", line 427, in dto_extractor
    return data_dto(connection).decode_bytes(body)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/dto/base_dto.py", line 96, in decode_bytes
    return backend.populate_data_from_raw(value, self.asgi_connection)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/dto/_backend.py", line 299, in populate_data_from_raw
    source_data=self.parse_raw(raw, asgi_connection),
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/dto/_backend.py", line 206, in parse_raw
    result = decode_json(value=raw, target_type=self.annotation, type_decoders=type_decoders)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/serialization/msgspec_hooks.py", line 191, in decode_json
    raise SerializationException(str(msgspec_error)) from msgspec_error
litestar.exceptions.base_exceptions.SerializationException: decoding to str: need a bytes-like object, type found - at `$.email`

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/middleware/exceptions/middleware.py", line 191, in __call__
    await self.app(scope, receive, send)
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 79, in handle
    response = await self._get_response_for_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 131, in _get_response_for_request
    response = await self._call_handler_function(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 160, in _call_handler_function
    response_data, cleanup_group = await self._get_response_data(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 186, in _get_response_data
    raise ClientException(str(e)) from e
litestar.exceptions.http_exceptions.ClientException: 400: decoding to str: need a bytes-like object, type found - at `$.email`
Traceback (most recent call last):
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/serialization/msgspec_hooks.py", line 113, in default_deserializer
    return decoder(target_type, value)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: decoding to str: need a bytes-like object, type found

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/serialization/msgspec_hooks.py", line 187, in decode_json
    return msgspec.json.decode(
           ^^^^^^^^^^^^^^^^^^^^
msgspec.ValidationError: decoding to str: need a bytes-like object, type found - at `$.email`

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 184, in _get_response_data
    kwargs["data"] = await kwargs["data"]
                     ^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/_kwargs/extractors.py", line 427, in dto_extractor
    return data_dto(connection).decode_bytes(body)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/dto/base_dto.py", line 96, in decode_bytes
    return backend.populate_data_from_raw(value, self.asgi_connection)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/dto/_backend.py", line 299, in populate_data_from_raw
    source_data=self.parse_raw(raw, asgi_connection),
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/dto/_backend.py", line 206, in parse_raw
    result = decode_json(value=raw, target_type=self.annotation, type_decoders=type_decoders)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/serialization/msgspec_hooks.py", line 191, in decode_json
    raise SerializationException(str(msgspec_error)) from msgspec_error
litestar.exceptions.base_exceptions.SerializationException: decoding to str: need a bytes-like object, type found - at `$.email`

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/middleware/exceptions/middleware.py", line 191, in __call__
    await self.app(scope, receive, send)
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 79, in handle
    response = await self._get_response_for_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 131, in _get_response_for_request
    response = await self._call_handler_function(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 160, in _call_handler_function
    response_data, cleanup_group = await self._get_response_data(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/geeshta/prog/litestar/pydbug/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 186, in _get_response_data
    raise ClientException(str(e)) from e
litestar.exceptions.http_exceptions.ClientException: 400: decoding to str: need a bytes-like object, type found - at `$.email`
```
</details>
### Litestar Version

2.0.1

### Platform

- [X] Linux
- [ ] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---

## Funding
* If you would like to see an issue prioritized, make a pledge towards it!
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/litestar-org/litestar/issues/2334">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2334/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by jw-can**
### Pull Request Checklist

- [x] New code has 100% test coverage
- [x] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [x] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [x] Pre-Commit Checks were ran and passed
- [x] Tests were ran and passed

### Description

- downtype Pydantic types which cannot be instantiated (e.g. EmailStr)

### Close Issue(s)

- Fixes #2334

### CODE REVIEW

File: tests/unit/test_contrib/test_pydantic/test_integration.py [hunk 1]
```diff
@@ -126,3 +126,18 @@ def test(
                     "key": "other_child.val.1",
                 },
             ]
+
+
+def test_pydantic_correct_emailstr_succeeds() -> None:
```
Comment (peterschutt) [inline tests/unit/test_contrib/test_pydantic/test_integration.py:131]: In addition to the integration tests, we should unit test `PydanticDTOFactory.generate_field_definitions()` using a model with types that we expect to downcast as input.

**Review by peterschutt — CHANGES_REQUESTED** [review]
Thanks for this!

A couple of comments.

I did some experimentation on types like `Json` earlier, and have made a PR into your fork with some suggestions based on that.

**Comment by JacobCoffee** [comment]
@jw-can anything to update here?

**Comment by jw-can** [comment]
Hi @JacobCoffee 
I've been super busy the last weeks and unfortunately couldn't get to work on the forks I prepared for litestar.
This one's hanging on Pydantic tests but as far as I can remember functionally everything's there in this fork for the issue.

**Comment by bmle69** [comment]
Just chiming in here after looking to see if an issue already exists for this problem. Seeing the same for url fields using pydantic models for the URL field

``{
  "status_code": 400,
  "detail": "Unsupported type: <class 'str'> - at `$.website_url`"
}``

**Comment by peterschutt** [comment]
Closed in favor of #3296 - thanks for taking a shot at this @jw-can. I've decided to take a more deliberate approach than relying on the openapi type.

**Pull request closed without merging**